### ISSUE TITLE
Global resource limits 

### ISSUE THREAD

**Issue by ozahavi** [aws/karpenter-core#745]
**Tell us about your request**
The ability to configure resource limits (CPU\GPU) on a global level that will be enforced\applied to all provisioners in the cluster.

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
In certain use cases, I need to enable a mix of several CPU and GPU instance types that are configured in separate provisioners (because of different taints that are added to the nodes). Without a global limit for resources, I cannot limit the overall capacity of nodes created by Karpenter in the cluster.

**Are you currently working around this issue?**
No.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by stevehipwell** [comment]
@njtran have there been any more discussions about this? My understanding is that part of the Karpenter v1 process is to achieve parity with CA and this functionality would be required for that? This is also a significant feature in the context of multi-role cluster operations where there should be central controls in place to support constraining self-service capabilities.

**Comment by jonathan-innis** [comment]
> My understanding is that part of the Karpenter v1 process is to achieve parity with CA 

I'm not sure that this is 100% correct here. Our goal here isn't necessarily to have 100% parity with CA but to provide a complete set of features and stable APIs that we believe are sufficient for all users to handle their node provisioning and lifecycle handling. Whether there is cross-over with CA or not is somewhat immaterial to whether we decide to scope a feature into GA for Karpenter v1.

> This is also a significant feature in the context of multi-role cluster operations where there should be central controls in place to support constraining self-service capabilities

As for multi-role cluster operations, I'm curious how you are thinking about how the self-service aspect of this would work, considering NodePools aren't namespaced at all today. If you have a user that has permission to update a NodePool, then you most likely also have permission to edit or delete other NodePools. Are you thinking that you would have some sort of policy engine in place that would allow you to determine whether someone does or doesn't have permission to do certain things on a certain set of NodePools? 

In either case, I could see a policy engine-based solution where you could reject a NodePool being updated or added if you took the sum of all the limits values on various NodePools and then saw that the sum of those limits is beyond what you would allow. In either case, you probably want a selector or namespaced-based approach to sum these limits across these NodePools.

**Comment by stevehipwell** [comment]
For the multi-role implementation, I'd expect the cluster owner to be able to configure constraints that even users with node pool rights couldn't override; much like how the CA model works today. On top of that I'd expect to layer admission control constraints to manage workloads and pools, but that's not really important in this context.

The most basic model here is for a cluster to be provisioned with a number of complimentary pools and then an upper resource limit to be made up from any combination of nodes.

**Comment by jonathan-innis** [comment]
I guess I'm trying to understand the utility of that. It's not self-service in the sense that you would have a multi-tenant cluster. It's self-service tenancy so you pretty much own the cluster except for the actual stand-up of the cluster. So someone comes in and pre-creates the EKS cluster and pre-installs Karpenter and let you have your way with what kind of instances you can launch. Did I get that right?

I'm also curious if there are AWS control-plane controls over resource totals associated with a particular tag or particular user agent. 

There may be ways to solve this at two different levels.

**Comment by stevehipwell** [comment]
@jonathan-innis yes that is the basic pattern. We're always running multiple pools and some of these overlap, so the current pool level limits aren't sufficient.

In the context of a single cloud we might be able to use cloud controls for this, but as Karpenter is multi -cloud I think this functionality should be Karpenter native.

**Comment by nclaeys** [comment]
I have another usecase for using nodes instead of cpu,memory limits within a cluster. We lately had a big outage at one of our customers because of a bigger than expected increase in nodes (500 -> 800). 
The memory requirements for many kubernetes operators (Karpenter, our own operator, cilium) are correlated to the number of nodes. A big increase in nodes caused several of these operators to crash repeatedly, although not enough to be in crashloop backoff. 
In our case Karpenter crashed after 1-2 minutes each time, which is why it did not scale down our expensive gpu nodepools, which were kept running for a long time even though they were not used. 

As a quick way of preventing this, we would love to limit the maximum number of nodes in cluster. This way we can guarantee that the settings we use will work up to a size that we tested for. If it goes above, we will get alerts but no hidden failures.

As we use 15 different types of nodepools, we do not want to set limits on each one. Also our situation does not care about which nodepool created the nodes, such that a limit on every node pool would lead to many inefficiencies.

Is there a specific reason why you do not consider global limits on a cluster? I think it is very sensible to make sure the cluster will behave as tested/expected. I rather have correct behavior but some jobs running slower than that all jobs on the cluster are affected.
